packages/app-content-pages/Dockerfile

```diff
@@ -1,9 +1,21 @@
 FROM zooniverse/front-end-monorepo:latest
```

Comment:
If I'm testing the NextJS app locally, is there a way that I can check which versions of the shared libraries are installed with this image?

For instance, can I use this base image for local development if I'm working on a PR that also adds new components to the zooniverse react components library? I'm not 100% clear whether this uses the local symlinks for installed dependencies, or whether it forces me to use the libs that were built during the most recent Jenkins build and pushed to `zooniverse/front-end-monorepo:latest`.

Reply:
The easiest way to dev / test local builds is to use the project root dockerfile and compose
as it will rebuild a local image with the local deps and the project and content-pages app. 

Local deps will be based on the local bootstrap image and will including the bootstrapped local react components and lib classifier code as well.

Reply:
The top-level Dockerfile bootstraps the monorepo and runs production builds of the NextJS apps (`yarn start`). For local dev work, I want to use a dev build of the NextJS app ie. `yarn dev` so that I get hot reloading for code changes.

Reply:
Right - so in the root compose file we currently call `yarn start`, we could switch that to use 'yarn dev' and inject the local package code into the running image for hot code reloading etc. Would that help with local testing etc?

FWIW the per package docker-compose files now use `yarn dev` and inject the local code to the container for hot code loading etc.
- https://github.com/zooniverse/front-end-monorepo/blob/27c3dfc186810f3fd112b4ca75352fef41350a79/packages/app-project/docker-compose.yml#L8
- https://github.com/zooniverse/front-end-monorepo/blob/27c3dfc186810f3fd112b4ca75352fef41350a79/packages/app-content-pages/docker-compose.yml#L22

ok, i think there is some confusion about my intentions with this PR. I'm trying to ensure our CI tools inject the correct build ENV data for the resulting production images.

I think the issue here is that our CI tool builds a bootstrap base image from the root Docker file (the rest of the multi stage docker file data is ignored). During CI we restrict concurrent builds to ensure the `zooniverse/front-end-monorepo:latest` image reflects our previous CI build step and then build both the project-app and content-pages package apps from their package docker files, these are isolated builds that we use in production. 

This CI process means that our root compose and dockerfiles aren't reproducing the production builds as they stand and most likely shouldn't be trying too either.

Reply:
>Local deps will be based on the local bootstrap image and will including the bootstrapped local react components and lib classifier code as well.

If we're using local builds and dependencies, why is the build also pulling an image of the full monorepo down from dockerhub?

To test this out, I added the sentence "This is a local build" to the classifier code and rebuilt with `yarn build`.

Running `yarn dev` in `packages/app-project`, here's what I see for http://localhost.zooniverse.org:3000/projects/nora-dot-eisner/planet-hunters-tess/classify?env=production "This is a local build" shows, as expected.
![The Planet Hunters TESS classifier with the words "This is a local build." above the subject viewer.](https://user-images.githubusercontent.com/59547/74743402-39e30380-5258-11ea-86c9-a1d1fe1b2cf1.png)

Shutting that down, then building and starting the project app with `docker-compose`, here's what I see. "This is a local build" doesn't show because I'm using the most recent copy of the classifier from the Jenkins build. This makes local dev unpredictable, depending on whether the most recent build was a staging or production build.
![The Planet Hunters TESS classifier looking exactly as it does on the live web site.](https://user-images.githubusercontent.com/59547/74743544-7878be00-5258-11ea-980c-f44e9f20984b.png)

Reply:
Again - this is a difference of CI / production builds using the freshly minted `zooniverse/front-end-monorepo:latest` image which doesn't have your code. We need this isolation for production builds in the per package Dockerfile. 

We can support your use case for local development work. My current train of thought is to provide that via the root dockerfile / compose files vs the per package ones. I chose the per package compose setup to prioritise testing of our production image builds not local dev work, we can change this.

Reply:
>Right - so in the root compose file we currently call yarn start, we could switch that to use 'yarn dev' and inject the local package code into the running image for hot code reloading etc. Would that help with local testing etc?

Not really, because we do also want that local production build of everything as a final check of any code changes. We have encountered bugs in the past that didn't show up when running `yarn dev` but did stop the project and content apps working when run with `yarn build && yarn start`.

Our current workflow is to run dev servers && apps in the local package directories, while changing code, then run a production build at the top level as a final check when we're happy with our feature branch.

Reply:
ok then. We have to decide how to test the production image changes for CI / deployment. Currently each package's Dockerfile is not used apart from CI builds and deployments. They are not used for local testing anywhere else.

This uses the root dockerfile not the package one, https://github.com/zooniverse/front-end-monorepo/blob/2234b9cae3f0fc967d430e1016b8636564b8a083/docker-compose.yml#L15

The per package docker-compose also does not use the package dockerfile, it uses the root dockerfile https://github.com/zooniverse/front-end-monorepo/blob/2234b9cae3f0fc967d430e1016b8636564b8a083/packages/app-project/docker-compose.yml#L7

This is not good, we really should be ensuring our deployment dockerfiles are tested and working as we expect. Hence my changes in the package docker-compose files to support this. 

We need to ensure we support both ensuring our production images work as expected (and are tested) and local development workflows and testing. I think we need two root compose files:
- One for production builds and testing nextjs app via `yarn start` on each package
- One for local code development workflows via `yarn dev` on each package

Reply:
Alternatively we can use a different per package testing setup (dev dockerfile / different compose setup) to ensure the local changes across apps are included in the setup. This would look similar to the current setup. I can switch the current compose setup for production testing explicitly. I'm happy to do this.

However we need to ensure we test both cases and ensure our deployment Dockerfiles are correctly setup, working as expected and we can test them locally.

Reply:
To be honest, I would add the NextJS app builds back into the `bootstap.sh` script, then use one bootstrap image for everything. I think that's the simplest solution.

Reply:
it still won't test the production dockerfile changes that i'm proposing in this PR, we need a way to test them.

Reply:
If the base image builds the production JS files for the Next JS apps, we can get rid of the individual app images. Will we still need the per-app Dockerfiles to run "yarn start" when the base image is deployed to kubernetes?

Reply:
doing so will increase the bloat of each docker image (i.e. the project-app will have the code  for the content-pages app, etc).

> Will we still need the per-app Dockerfiles to run "yarn start" when the base image is deployed to kubernetes?
Not necessarily, we can use args via K8s to run the particular app code, https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/#define-a-command-and-arguments-when-you-create-a-pod

I still don't think that's a great solution, i'm working on something that might alleviate the issues with dev / deployment workflows.

Reply:
>doing so will increase the bloat of each docker image (i.e. the project-app will have the code for the content-pages app, etc).

That's not a big deal, to be honest, when you consider that each app image still has all the source code for the libraries etc.

Reply:
Something like this might help with allowing normally app testing as you've outlined and provide a way to test the production docker images / builds.
https://github.com/zooniverse/front-end-monorepo/pull/1522

Reply:
👍 I'll take a look.

Building a single base image would also have the advantage that we can tag it with a commit ID. The more I think about that idea, the more I like it.

Reply:
It might simplify the setup, we will still need a root and per package compose file for booting the apps locally for dev / testing workflows.

the single dockerfile works till we have different env vars etc, at that point we could re-split the root dockerfile up to something like we've got now.

Reply:
Ok - based on the latest changes from #1522 are you happy with your workflow for testing etc with the way this is now setup?